attachment_preview/model/ir_attachment.py

```diff
@@ -32,35 +32,50 @@ def get_binary_extension(
-            if not this[binary_field]:
-                result[this.id] = False
-                continue
-            if not extension:
-                try:
-                    import magic
-                    ms = magic.open(
-                        hasattr(magic, 'MAGIC_MIME_TYPE') and
-                        magic.MAGIC_MIME_TYPE or magic.MAGIC_MIME)
-                    ms.load()
+        context_bin_size = dict(context or {})
+        context_bin_size['bin_size'] = True
+        ids_to_browse = ids if isinstance(ids, collections.Iterable) \
+            else [ids]
+        # First pass: load fields in bin_size mode to avoid loading big files
+        #  unnecessarily.
+        if filename_field:
+            for this in self.pool[model].browse(cr, uid, ids_to_browse,
+                                                context=context_bin_size):
+                if not this.id:
+                    result[this.id] = False
+                    continue
+                extension = ''
+                if this[filename_field]:
+                    filename, extension = os.path.splitext(
+                        this[filename_field])
+                if this[binary_field] and extension:
+                    result[this.id] = extension
+        # Second pass for all attachments which have to be loaded fully
+        #  to get the extension from the content
+        ids_to_browse = [_id for _id in ids_to_browse if _id not in result]
+        for this in self.pool[model].browse(cr, uid, ids_to_browse,
+                                            context=context):
+            try:
+                import magic
+                ms = magic.open(
+                    hasattr(magic, 'MAGIC_MIME_TYPE') and
+                    magic.MAGIC_MIME_TYPE or magic.MAGIC_MIME)
+                ms.load()
+                if model == self._name and binary_field == 'datas'\
+                        and this.store_fname:
+                    mimetype = ms.file(
+                        this._full_path(cr, uid, this.store_fname))
+                else:
                     mimetype = ms.buffer(
                         base64.b64decode(this[binary_field]))
-                except ImportError:
-                    (mimetype, encoding) = mimetypes.guess_type(
-                        'data:;base64,' + this[binary_field], strict=False)
-                extension = mimetypes.guess_extension(
-                    mimetype.split(';')[0], strict=False)
-
-            result[this.id] = (extension or '').lstrip('.').lower()
+            except ImportError:
+                (mimetype, encoding) = mimetypes.guess_type(
+                    'data:;base64,' + this[binary_field], strict=False)
+            extension = mimetypes.guess_extension(
+                mimetype.split(';')[0], strict=False)
+            result[this.id] = extension
+        for _id in result:
+            result[_id] = (extension or '').lstrip('.').lower()
```

Comment:
was this here before? This won't crash most of the time, but assign the last extension we got to all requested records.

Maybe add a test for this function so that we're sure it behaves correctly?